clippy_lints/src/bool_to_int_with_if.rs

```diff
@@ -0,0 +1,123 @@
+use rustc_ast::{ExprPrecedence, LitKind};
+use rustc_hir::{Block, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+use clippy_utils::{diagnostics::span_lint_and_then, source::snippet_block};
+use rustc_errors::Applicability;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Instead of using an if statement to convert a bool to an int, this lint suggests using a `as` coercion or a `from()` function.
+    /// ### Why is this bad?
+    /// Coercion or `from()` is idiomatic way to convert bool to a number.
+    /// Both methods are guaranteed to return 1 for true, and 0 for false. See https://doc.rust-lang.org/std/primitive.bool.html#impl-From%3Cbool%3E
+    ///
+    /// ### Example
+    /// ```rust
+    /// let condition = false;
+    /// if condition {
+    ///     1_i64
+    /// } else {
+    ///     0
+    /// };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let condition = false;
+    /// i64::from(condition);
+    /// ```
+    /// or
+    /// ```rust
+    /// let condition = false;
+    /// condition as i64;
+    /// ```
+    #[clippy::version = "1.64.0"]
+    pub BOOL_TO_INT_WITH_IF,
+    style,
+    "using if to convert bool to int"
+}
+declare_lint_pass!(BoolToIntWithIf => [BOOL_TO_INT_WITH_IF]);
+
+impl<'tcx> LateLintPass<'tcx> for BoolToIntWithIf {
+    fn check_expr(&mut self, ctx: &LateContext<'tcx>, expr: &'tcx rustc_hir::Expr<'tcx>) {
+        if !expr.span.from_expansion() {
+            check_if_else(ctx, expr);
+        }
+    }
+}
+
+fn check_if_else<'tcx>(ctx: &LateContext<'tcx>, expr: &'tcx rustc_hir::Expr<'tcx>) {
+    if_chain!(
+        if let ExprKind::If(check, then, Some(else_)) = expr.kind;
+
+        if let Some(then_lit) = int_literal(then);
+        if let Some(else_lit) = int_literal(else_);
+
+        if check_int_literal_equals_val(then_lit, 1);
+        if check_int_literal_equals_val(else_lit, 0);
+
+        then {
+            let ty = ctx.typeck_results().expr_ty(then_lit); // then and else must be of same type
+
+            let need_parens = should_have_parentheses(check);
+            let snippet = snippet_block(ctx, check.span, "..", None);
```

Comment:
There are cases, where these snippet functions can't get the correct snippet, in those cases the placeholder is returned. Can you use [`snippet_block_with_applicability`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/source/fn.snippet_block_with_applicability.html) to ensure that the applicability is also adjusted correctly? The code would looks something like this:

```rs
			let mut app = Applicability::MachineApplicable;
            let snippet = snippet_block_with_applicability(ctx, check.span, "..", None, &mut app);
```

Then you can later use the applicability in `diag.span_suggestion(...)`

Reply:
fixed in [85b9c05](https://github.com/rust-lang/rust-clippy/pull/9086/commits/85b9c053154a449193b6e3dff9e3fcb35365c5f3)